src/py/flwr/client/process/clientappio_servicer.py

```diff
@@ -0,0 +1,111 @@
+# Copyright 2024 Flower Labs GmbH. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""ClientAppIo API servicer."""
+
+
+from logging import DEBUG, ERROR
+from typing import Tuple
+
+import grpc
+
+from flwr.common import Context, Message, typing
+from flwr.common.logger import log
+from flwr.common.serde import (
+    clientappstatus_to_proto,
+    context_from_proto,
+    context_to_proto,
+    message_from_proto,
+    message_to_proto,
+    run_to_proto,
+)
+from flwr.common.typing import Run
+
+# pylint: disable=E0611
+from flwr.proto import clientappio_pb2_grpc
+from flwr.proto.clientappio_pb2 import (  # pylint: disable=E0401
+    PullClientAppInputsRequest,
+    PullClientAppInputsResponse,
+    PushClientAppOutputsRequest,
+    PushClientAppOutputsResponse,
+)
+from flwr.proto.message_pb2 import Context as ProtoContext
+from flwr.proto.message_pb2 import Message as ProtoMessage
+from flwr.proto.run_pb2 import Run as ProtoRun
+
+
+# pylint: disable=C0103,W0613,W0201
+class ClientAppIoServicer(clientappio_pb2_grpc.ClientAppIoServicer):
+    """ClientAppIo API servicer."""
+
+    def set_object(
+        self,
+        message: Message,
+        context: Context,
+        run: Run,
+        token: int,
+    ) -> None:
+        """Set client app objects."""
+        log(DEBUG, "ClientAppIo.SetObject")
+        # Serialize Message, Context, and Run
+        self.proto_message: ProtoMessage = message_to_proto(message)
+        self.proto_context: ProtoContext = context_to_proto(context)
+        self.proto_run: ProtoRun = run_to_proto(run)
+        self.token: int = token
+
+    def get_object(self) -> Tuple[Message, Context]:
+        """Get client app objects."""
+        log(DEBUG, "ClientAppIo.GetObject")
+        return self.message, self.context
+
+    def _update_object(self) -> None:
+        """Update client app objects."""
+        log(DEBUG, "ClientAppIo.UpdateObject")
+        # Deserialize Message and Context
+        self.message: Message = message_from_proto(self.proto_message)
+        self.context: Context = context_from_proto(self.proto_context)
```

Comment:
wouldn't it be better to define these elements during `__init__()`?

Reply:
Unfortunately not. This instance of `ClientAppIoServicer` is created when the `grpc.Server` is spun up for the first time and there is no `Message` or `Context` at that point (see [this line](https://github.com/adap/flower/blob/808c2fbdb3300ec05f6a8d454dbb8a13aa5e69cf/src/py/flwr/client/app.py#L680)). I agree it's not elegant, do you know of a better solution @jafermarq?

Reply:
I was thinking on just define them like this?
```python
def __init__(self, ...):
    self.message: Optional[Message] = None
    self.context: Optional[Context] = None
```

Reply:
This was my first implementation, but I got a type mismatch error during the linting CI, which I couldn't work around.

Reply:
i see. One solution is to  put those `*_from_proto()` behind a `if not None` (that should keep mypy happy)